bip-0077.md

```diff
@@ -0,0 +1,703 @@
+third parties.
+
+### Uniform Payloads
+
+Encapsulated OHTTP payloads seen by the relay and directory, and
+encrypted messages seen by the directory, are constructed to be uniform
+so that these third-party services are unable to distinguish between
+them.
+
+Encapsulated OHTTP messages are 8192 bytes long, and begin with a
+cleartext OHTTP header and an uncompressed key which is distinguishable
+from random bytes but uniform across different encapsulated requests.
+
+End-to-end encrypted messages are 7168 bytes long, and should be
+indistinguishable from uniformly random bytes.
+[ElligatorSwift as defined in BIP 324](https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki#elligatorswift-encoding-of-curve-x-coordinates)
+is used to encode encapsulated HPKE public keys prepended to the HPKE ciphertext
+so that the directory can't distinguish between key material, the
+ciphertext, and randomness. This ensures the two different protocol
+messages are indistinguishable from each other as well as any protocol
+extensions.
+
+These padded sizes are sufficient for most PSBTs without exceeding the [
+8KB
+limit](https://www.geekersdigest.com/max-http-request-header-size-server-comparison/)
+of many HTTP/1.1 web servers. 8KB is also too small for image sharing,
+making misuse of the directory impractical.
+
+### Secp256k1 Hybrid Public Key Encryption
+
+[RFC 9180 Hybrid Public Key
+Encryption](https://www.rfc-editor.org/rfc/rfc9180.html)
+(HPKE) is a modern IETF standard for secure
+message exchange without TLS, which is not available in Bitcoin Core.
+
+Payjoin version 2 uses `DHKEM(Secp256k1, HKDF-SHA256)` and
+`ChaCha20Poly1305` AEAD for both OHTTP encapsulation and for end-to-end
+encryption between the sender and receiver.
+
+The receiver transmits its receiver key in [receiver fragment
+parameters](#receiver-fragment-parameters). The sender shares
+its reply key along with the Original PSBT. These keys are ephemeral and
+must only be used for a single Payjoin Session.
+
+#### Secp256k1-based DHKEM
+
+[Secp256k1-based DHKEM for
+HPKE](https://www.ietf.org/archive/id/draft-wahby-cfrg-hpke-kem-secp256k1-01.html)
+is most appropriate because of secp256k1's availability in bitcoin
+contexts.
+
+#### ChaCha20Poly1305 AEAD
+
+This authenticated encryption with additional data [
+algorithm](https://en.wikipedia.org/wiki/ChaCha20-Poly1305)
+is standardized in [RFC
+8439](https://www.rfc-editor.org/rfc/rfc8439) and has high
+performance. ChaCha20Poly1305 AEAD has been implemented [in Bitcoin
+Core](https://github.com/bitcoin/bitcoin/pull/15649) for [
+BIP 324 Encrypted
+Transport](https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki)
+as well. This has widespread support in browsers and common
+cryptographic libraries. AES-GCM is more widespread but slower without
+hardware support and not typically already a dependency in bitcoin software.
+
+#### HKDF-SHA256
+
+SHA-256 is necessarily available in bitcoin contexts.
```

Comment:
There may be a need for more information about HKDF-SHA256.
Bitcoin Core, for example, has an HKDF-SHA256 implementation, but it only supports an output length of 32 bytes (L=32).
https://github.com/bitcoin/bitcoin/blob/746ab19d5a13c98ae7492f9b6fb7bd6a2103c65d/src/crypto/hkdf_sha256_32.h#L13

Is this implementation, for example, sufficient for Payjoin, or do we need a fully RFC 5896-compliant HKDF-SHA256 implementation?

Reply:
The rationale for choosing HKDF-SHA256 is not that it is already implemented in bitcoin core, but that its primitive, sha-256, is necessarily available in all contexts.

However, the linked [secp256k1-based DHKEM for HPKE draft](https://www.ietf.org/archive/id/draft-wahby-cfrg-hpke-kem-secp256k1-01.html) explicitly sets Nsecret = 32 for the secp256k1 KEM, so Bitcoin Core’s L=32-constrained HKDF‑SHA256 should satisfy spec out of the box.